python/lsst/daf/butler/tests/_testRepo.py

```diff
@@ -370,45 +453,33 @@ def addDataIdValue(butler: Butler, dimension: str, value: Union[str, int], **rel
-            matches = list(
-                butler.registry.queryDimensionRecords(dimension_obj, dataId=match_data_id).limit(1)
-            )
-            if not matches:
-                # Nothing in the registry matches: invent a data ID value
-                # with the right type (actual value does not matter).
-                dimension_value = _makeRandomDataIdValue(dimension_obj)
-                data_id[dimension_obj.name] = dimension_value
-            else:
-                # A record does exist in the registry.  Use its data ID value,
-                # and skip the rest of the record-fabrication logic.
-                data_id[dimension_obj.name] = matches[0].dataId[dimension_obj.name]
+            data_id[dimension_obj.name], invented = _findOrInventDataIdValue(butler, data_id, dimension_obj)
+            if not invented:
+                # No need to make a new record; one already exists.
                 continue
+        if dimension_obj.name in related:
+            # Caller passed in a value of this dimension explicitly, but it
+            # isn't the primary dimension they asked to have a record created
+            # for.  That means they expect this record to already exist.
+            continue
+        if dimension_obj != fullDimension and dimension_obj in all_dimensions.required:
+            # We also don't want to automatically create new dimension records
+            # for required dimensions (except for the main dimension the caller
+            # asked for); those are also asserted by the caller to already
+            # exist.
+            continue
         if dimension_obj.viewOf is not None:
             # Don't need to bother generating full records for dimensions whose
             # records are just a view into some other's records anyway.
             continue
-        # Add the primary key field for this dimension.
-        record_dict: dict[str, Any] = {dimension_obj.primaryKey.name: dimension_value}
-        # Define secondary keys (e.g., detector name given detector id)
-        record_dict.update(_fillAllKeys(dimension_obj, dimension_value))
-        # Set the foreign key values for any related dimensions that should
-        # appear in the record.
-        for related_dimension in dimension_obj.dimensions:
-            if related_dimension.name != dimension_obj.name:
-                record_dict[related_dimension.name] = data_id[related_dimension.name]
-
-        record_dicts_by_dimension[dimension_obj] = record_dict
+        record_dicts_by_dimension[dimension_obj] = _makeDimensionRecordDict(data_id, dimension_obj)
 
     # Sync those dimension record dictionaries with the database.
     for dimension_obj, record_dict in record_dicts_by_dimension.items():
         record = dimension_obj.RecordClass(**record_dict)
         try:
             butler.registry.syncDimensionData(dimension_obj, record)
-        except ConflictingDefinitionError as e:
+        except sqlalchemy.exc.IntegrityError as e:
```

Comment:
I don't understand this change. Was it a mis-specification of `syncDimensionData`?

Reply:
Short version: this is reverting the exception type back to `sqlalchemy.exc.IntegrityError`, which it was originally and probably should have been throughout my work on the branch, but in the middle I got confused about what the likely failure mode was.

Long version: `IntegrityError` is the right type for low-level foreign key and unique constraints (in this context, foreign key violations are much more likely, and what the re-raise here anticipates, but sqlalchemy makes it really hard to tell which, which is the main reason we don't re-raise at a lower level).  The changes on this branch that have not been dropped made that kind of constraint violation impossible, but I wasn't able to convince myself that a different error condition could be ruled out, in which the record already exists but it has different content, and `ConflictingDefinitionError` is raised.  So I switched the except block to that.  With the final version of the logic here that's more like the original, foreign key violations (due to missing required dependencies) are once again possible, and the original re-raise makes sense again.